docs/designs/restore-and-gc.md

```diff
@@ -0,0 +1,160 @@
+  - Provide a mechanism to lock down the version of dependency file/repository for different `build`s, just like [npm package locks](https://docs.npmjs.com/files/package-locks)
+
+## Restore files
+
+There are some cases that the docfx `build` depends on external files for help. 
+
+For example, if you want the docfx to resolve `gitHub contributors` from building file git commit history, the [GitHub User Cache](github-user-cache.md) file can be set to reduce the API calls to [GitHub](https://github.com). The file can be just a http(s) url, which `restore` will help to download for `build` usage.
+
+```yml
+github:
+  userCache: https:///some-blob-service/github-user-cache.json
+```
+
+The restored files will be stored at `%DOCFX_APPDATA_PATH%/downloads/{url-short-path}/{version}+{etag}` like:
+
+```text
+`%DOCFX_APPDATA_PATH%`
+    | - downloads
+        | - raw.gith..tent.com+docascode+docfx-te..ndencies+62b0448+extend1.yml+dc363b0e
+            | - 18b265bc4e9a6b4fdff8b5210caf9e3541f4cb2d+%22f02fde5b4f3817266458170ba048c7fc2de46287%22
+            | - 3111eb522b092c9cef5cd0ccd86892cf847592d6+%2299a5dbb8d73da1ec452d74c0ab45b11ebe8615a0%22
+
+```
+
+- `{url-short-path}` is calculated from `file` url
+- `{version}` is sha1 hash of file content, 
+- `{etag}` is timestamp or something else supported by the service which stores the `file`.
+
+If the `etag` is supported by its service, docfx avoids duplicated downloads for the file never changed.
+
+## Restore dependency repositories
+
+There are some cases that the docfx `build` depends on external repositories, which includes `tokens`, `code-snippets` or other resources/content.
+
+For example, some files are referencing `code example` for better explanations, but some of the code examples stores in another repository, maintained by another team/owners, with the dependency repository, you have the ability to reference external `code example` to be built in your page.
+
+Config(ocs/designs/config.md):
+
+``` yml
+dependencies:
+  docfx-code: https://github.com/dotnet/docfx#master
+```
+
+Content:
+
+```md
+Example: [!code-csharp[Main](docfx-code/program.cs)]
+```
+
+The restored dependency repositories will be stored at `%DOCFX_APPDATA_PATH%/git/{url-short-path}/{worktree-name}` organized by [git work tree](https://git-scm.com/docs/git-worktree) like:
+
+```text
+`%DOCFX_APPDATA_PATH%`
+    | - git
+        | - github.com+Microsoft+Luis-Samples+5e455995
+            | - .git
+            | - master-fd6968dec9a9a39aec1845232466fe35fca520da
+            | - live-a82cb2a51f6925b2fa87275218b01678c5aadfe1
+```
+
+- `{url-short-path}` is calculated from `git remote` url
+- `{worktree-name}` is combined by `branch name` and `HEAD commit id`
+
+## Dependency lock
+
+There are some cases you want to keep using the **same version** of dependency files/repository between different `build`s.
```

Comment:
This works for git, but not for file. We can go to a specific commit for a git repo, but cannot ask server to return a file with a specific e-tag existed before.

So, lock file can only consider git. DocFX always fetches latest for file dependency if cache miss.

Reply:
For files, could `Etag` work similar to `version`?

Reply:
already have NOTE about that

---
In reply to: [247030971](https://github.com/dotnet/docfx/pull/3900#discussion_r247030971) [](ancestors = 247030971)